lib/src/rules/unreachable_from_main.dart

```diff
@@ -0,0 +1,188 @@
+
+    var declarationByElement = <Element, Declaration>{};
+    for (var declaration in topDeclarations) {
+      var element = declaration.declaredElement;
+      if (element != null) {
+        if (element is TopLevelVariableElement) {
+          declarationByElement[element] = declaration;
+          var getter = element.getter;
+          if (getter != null) declarationByElement[getter] = declaration;
+          var setter = element.setter;
+          if (setter != null) declarationByElement[setter] = declaration;
+        } else {
+          declarationByElement[element] = declaration;
+        }
+      }
+    }
+
+    var dependencies = Map<Declaration, Set<Declaration>>.fromIterable(
+      topDeclarations,
+      value: (declaration) =>
+          DartTypeUtilities.traverseNodesInDFS(declaration as Declaration)
+              .expand((e) => [
+                    if (e is SimpleIdentifier) e.staticElement,
+                    // with `id++` staticElement of `id` is null
+                    if (e is CompoundAssignmentExpression) ...[
+                      e.readElement,
+                      e.writeElement,
+                    ],
+                  ])
+              .whereNotNull()
+              .map((e) {
+                while (e.enclosingElement2 != null &&
+                    e.enclosingElement2 is! CompilationUnitElement) {
+                  e = e.enclosingElement2!;
+                }
+                return e;
+              })
+              .map((e) => declarationByElement[e])
+              .whereNotNull()
+              .where((e) => e != declaration)
+              .toSet(),
+    );
+
+    var usedMembers = entryPoints.toSet();
+    var toTraverse = Queue.from(usedMembers);
+    while (toTraverse.isNotEmpty) {
+      var declaration = toTraverse.removeLast();
+      for (var dep in dependencies[declaration]!) {
+        if (usedMembers.add(dep)) {
+          toTraverse.add(dep);
+        }
+      }
+    }
+
+    var unusedMembers = topDeclarations.difference(usedMembers).where((e) {
+      var element = e.declaredElement;
+      return element != null &&
+          element.isPublic &&
+          !element.hasVisibleForTesting;
+    });
+
+    for (var member in unusedMembers) {
+      if (member is NamedCompilationUnitMember) {
+        rule.reportLint(member.name);
+      } else if (member is VariableDeclaration) {
+        rule.reportLint(member.name);
+      } else if (member is ExtensionDeclaration) {
+        rule.reportLintForToken(
+            member.name?.token ?? member.firstTokenAfterCommentAndMetadata);
+      } else {
+        rule.reportLintForToken(member.firstTokenAfterCommentAndMetadata);
+      }
+    }
+  }
+
+  bool _isEntryPoint(Declaration e) =>
+      e is FunctionDeclaration &&
+      (e.name.name == 'main' || e.metadata.any(_isPragmaVmEntry));
+
+  bool _isPragmaVmEntry(Annotation annotation) {
+    var elementAnnotation = annotation.elementAnnotation;
+    if (elementAnnotation != null) {
```

Comment:
5 levels of if-statements makes this code pretty hard for me to read. Can we instead short circuit `return false` when we hit a condition that makes this _not_ a pragma, and flatten out the code?